SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp

```diff
@@ -0,0 +1,202 @@
+
+	// Construct the mapping update based on the local virtual worker to physical worker mapping.
+	Worker_ComponentUpdate Update = {};
+	Update.component_id = SpatialConstants::VIRTUAL_WORKER_TRANSLATION_COMPONENT_ID;
+	Update.schema_type = Schema_CreateComponentUpdate();
+	Schema_Object* UpdateObject = Schema_GetComponentUpdateFields(Update.schema_type);
+
+	WriteMappingToSchema(UpdateObject);
+
+	check(Connection.IsValid());
+	Connection->SendComponentUpdate(SpatialConstants::INITIAL_VIRTUAL_WORKER_TRANSLATOR_ENTITY_ID, &Update);
+
+	// The Translator on the worker which hosts the manager won't get the component update notification,
+	// so send it across directly.
+	Translator->ApplyVirtualWorkerManagerData(UpdateObject);
+}
+
+void SpatialVirtualWorkerTranslationManager::QueryForWorkerEntities()
+{
+	UE_LOG(LogSpatialVirtualWorkerTranslationManager, Log, TEXT("Sending query for WorkerEntities"));
+
+	checkf(!bWorkerEntityQueryInFlight, TEXT("Trying to query for worker entities while a previous query is still in flight!"));
+
+	// Create a query for all the system entities which represent workers. This will be used
+	// to find physical workers which the virtual workers will map to.
+	Worker_ComponentConstraint WorkerEntityComponentConstraint{};
+	WorkerEntityComponentConstraint.component_id = SpatialConstants::WORKER_COMPONENT_ID;
+
+	Worker_Constraint WorkerEntityConstraint{};
+	WorkerEntityConstraint.constraint_type = WORKER_CONSTRAINT_TYPE_COMPONENT;
+	WorkerEntityConstraint.constraint.component_constraint = WorkerEntityComponentConstraint;
+
+	Worker_EntityQuery WorkerEntityQuery{};
+	WorkerEntityQuery.constraint = WorkerEntityConstraint;
+	WorkerEntityQuery.result_type = WORKER_RESULT_TYPE_SNAPSHOT;
+
+	// Make the query.
+	Worker_RequestId RequestID;
+	check(Connection.IsValid());
+	RequestID = Connection->SendEntityQueryRequest(&WorkerEntityQuery);
+	bWorkerEntityQueryInFlight = true;
+
+	// Register a method to handle the query response.
+	EntityQueryDelegate WorkerEntityQueryDelegate;
+	WorkerEntityQueryDelegate.BindRaw(this, &SpatialVirtualWorkerTranslationManager::WorkerEntityQueryDelegate);
+	check(Receiver.IsValid());
+	Receiver->AddEntityQueryDelegate(RequestID, WorkerEntityQueryDelegate);
+}
+
+// This method allows the translation manager to deal with the returned list of connection entities when they are received.
+// Note that this worker may have lost authority for the translation mapping in the meantime, so it's possible the
+// returned information will be thrown away.
+void SpatialVirtualWorkerTranslationManager::WorkerEntityQueryDelegate(const Worker_EntityQueryResponseOp& Op)
+{
+	bWorkerEntityQueryInFlight = false;
+
+	if (Op.status_code != WORKER_STATUS_CODE_SUCCESS)
+	{
+		UE_LOG(LogSpatialVirtualWorkerTranslationManager, Warning, TEXT("Could not find Worker Entities via entity query: %s, retrying."), UTF8_TO_TCHAR(Op.message));
```

Comment:
personal pref, I think this would be easier to follow if we had a call to `QueryForWorkerEntities();` followed by a `return;` then the rest of the function just assumes the query was successful

Reply:
Let me think about this a bit. I'm not sure we can't get into a situation where there's an extra query flying around.

Reply:
I think I'd rather leave it like this. The main advantage to this flow is that there's only one place where QueryForWorkerEntities is done. Otherwise we'd have two calls, and I personally find that a bit harder to reason about.